feat: 🎸 Add interpolation to configuration

Waterstraal opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/transloco/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

You cannot specify your own regular expression for matching parameters. 

Issue Number: #323 

## What is the new behavior?

A new transpiler has been added that adds the possibility to specify your own regular expression for matching parameters. 

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

shaharkazaz commented:
@Waterstraal Thanks for the PR!

I don't think this should be an actual transpiler, I believe that a property in the config is more fitting here.
Can you replace the transpiler with the `interpolation` config property and consume it in the `defaultTranspiler`? 

It should support the following values:
* A Regex
* An array that indicates the start and end markings ([start, end]: [string, string]).

A couple of things to keep in mind:
* Validate that the input.
* If a regex is provided we need to make sure the global flag is present.

Waterstraal commented:
> @Waterstraal Thanks for the PR!
> 
> I don't think this should be an actual transpiler, I believe that a property in the config is more fitting here.
> Can you replace the transpiler with the `interpolation` config property and consume it in the `defaultTranspiler`?
> 
> It should support the following values:
> 
> * A Regex
> * An array that indicates the start and end markings ([start, end]: [string, string]).
> 
> A couple of things to keep in mind:
> 
> * Validate that the input.
> * If a Regex is provided we need to make sure the global flag is present.

Good suggestion. I pushed a new commit that adds `interpolation` to Transloco config

shaharkazaz commented on docs/docs/transpiler.mdx:
```diff
@@ -6,6 +6,38 @@ description: The Transpiler | Transloco Angular i18n
 The transpiler is responsible for resolving the given value.  
 For example, when given `Hello {{ key }}` the default transpiler will replace the dynamic variable `key` based on the given params, or in some cases, within the [translation object itself](./additional-functionality#reference-other-keys).
 
+## DefaultTranspiler
```
The documentation for the new property should be part of the [config options page](https://ngneat.github.io/transloco/docs/config-options/).

shaharkazaz commented on projects/ngneat/transloco/src/lib/transloco.transpiler.ts:
```diff
@@ -72,6 +80,19 @@ export class DefaultTranspiler implements TranslocoTranspiler {
 }
 
+function getInterpolationRegExpFromConfig(userConfig?: TranslocoConfig): RegExp {
+  const interpolation = userConfig ? userConfig.interpolation : defaultConfig.interpolation;
```
Looking at the code now, I don't think we need to support Regex, don't see any reason for someone to pass a regex instead of the markings. Do you agree?

Update the default value to: `['{{', '}}']`
```
const [start, end] = userConfig?.interpolation ?? defaultConfig.interpolation;

return new RegExp(`${start}(.*?)${end}`, 'g');
```

Waterstraal replied:
Makes sense to me to remove the RegExp 👍  

Sadly we can't use optional chaining and nullish coalescing as this project is still on typescript 3.2

shaharkazaz commented on projects/ngneat/transloco/src/lib/transloco.transpiler.ts:
```diff
@@ -1,18 +1,26 @@
 }
 
 export class DefaultTranspiler implements TranslocoTranspiler {
+  protected interpolationRegExp: RegExp;
```
How about interpolationMatcheris?

shaharkazaz commented on projects/ngneat/transloco/src/lib/transloco.transpiler.ts:
```diff
@@ -72,6 +80,19 @@ export class DefaultTranspiler implements TranslocoTranspiler {
   }
 }
 
+function getInterpolationRegExpFromConfig(userConfig?: TranslocoConfig): RegExp {
```
How about `resolveMatcher`?

shaharkazaz commented on projects/ngneat/transloco/src/lib/tests/transloco.transpiler.spec.ts:
```diff
@@ -86,30 +108,42 @@ describe('TranslocoTranspiler', () => {
       const lang = flatten({
-        dear: 'dear {{name}}',
-        hello: 'Hello {{dear}}'
+        dear: `dear ${interpolationMarkings[0]}name${interpolationMarkings[1]}`,
```
create a function called `wrapParam` which replaces: `${interpolationMarkings[0]}paramName${interpolationMarkings[1]}` and replace all instances with the function call.

shaharkazaz commented on projects/ngneat/transloco/src/lib/tests/transloco.transpiler.spec.ts:
```diff
@@ -63,19 +68,27 @@ describe('TranslocoTranspiler', () => {
+  function testDefaultBehaviour(parser: TranslocoTranspiler, interpolationMarkings: [string, string] = ['{{', '}}']) {
     it('should translate simple string from params', () => {
-      const parsed = parser.transpile('Hello {{ value }}', { value: 'World' }, {});
+      const parsed = parser.transpile(`Hello ${wrapParam('value', interpolationMarkings)}`, { value: 'World' }, {});
```
why do you need to pass the interpolationMarkings to the function? just create it in the testDefaultBehaviour scope.

Waterstraal commented:
@shaharkazaz I was wondering whether we should make `TranslocoConfig` a required field in `DefaultTranspiler`'s constructor that all derived transpilers have to pass it in. This prevents a potential issue with a user configuration that specifies a custom `interpolation` in combination with a custom transpiler that doesn't pass the `TranslocoConfig` in. Your thoughts?

shaharkazaz commented:
@Waterstraal that would make it a breaking change and we don't want that 🙂  maybe in v3 😉 
Regarding the MessageFormatTranspiler, we need to pass the config there.

Waterstraal commented:
> @Waterstraal that would make it a breaking change and we don't want that 🙂 maybe in v3 😉
> 
> Regarding the MessageFormatTranspiler, we need to pass the config there, nice catch.

Shall I update `MessageFormatTranspiler`?

shaharkazaz commented:
@Waterstraal Yes, please test it in the playground and make sure everything is working as expected.

Waterstraal commented:
@shaharkazaz I updated the transpiler and added some specs and udpated cypress to verify it's working. I tested it manually on the playground by adding interpolation settings and updating message bundles. It's difficult in the current setup to have multiple instances of TranslocoConfig provided in the playground I think? Thant would be ideal to test multiple scenario's.
